### PR TITLE
Enabling gradient checkpointing in eval() mode

### PR DESCRIPTION
**Pull request by MikeTkachuk**
Removed unnecessary `if self.training ...` check when using gradient checkpointing
https://github.com/huggingface/diffusers/issues/9850

### CODE REVIEW

**Comment by MikeTkachuk** [comment]
Since all of the module implementations used the same `if self.training and self.gradient_checkpointing:` clause, I wonder if there is some reference template for `def forward` that others are using? Need to make sure future implementations are using the fixed clause

File: src/diffusers/models/transformers/cogvideox_transformer_3d.py [hunk 1]
```diff
@@ -452,7 +452,7 @@ def forward(
 
         # 3. Transformer blocks
         for i, block in enumerate(self.transformer_blocks):
-            if self.training and self.gradient_checkpointing:
+            if self.gradient_checkpointing:
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/cogvideox_transformer_3d.py:455]: oh thanks! why do we also removed the `torch.is_grad_enabled()` check? gradient checkpointing isn't meaningful without gradient being computed, no?

Comment (MikeTkachuk) [reply src/diffusers/models/transformers/cogvideox_transformer_3d.py]: added it back, thanks for pointing it out. 
it does not break anything, but found that it throws an annoying warning when use_reentrant=True,

Comment (yiyixuxu) [reply src/diffusers/models/transformers/cogvideox_transformer_3d.py]: >but found that it throws an annoying warning when use_reentrant=True,

what do you mean by that?

Comment (MikeTkachuk) [reply src/diffusers/models/transformers/cogvideox_transformer_3d.py]: use_reentrant is an argument passed to torch.utils.checkpoint.checkpoint

if True one of the checks will print this to stderr
warnings.warn(
            "None of the inputs have requires_grad=True. Gradients will be None"
        )
 but diffusers are using use_reentrant=False anyway

Comment (yiyixuxu) [reply src/diffusers/models/transformers/cogvideox_transformer_3d.py]: oh got thanks, so the warning is specific to when we use gradient checkpointing when gradient is not enabled

**Comment by yiyixuxu** [comment]
hi @MikeTkachuk unfortunately we have to update the branch and resolve conflicts now...
would you be able to do that? I'm cool with the change otherwise and can merge once it is synced with main

**Comment by MikeTkachuk** [comment]
done, also fixed in a few other places I missed

File: docs/source/en/api/models/controlnet.md [hunk 2]
```diff
@@ -39,12 +39,12 @@ pipe = StableDiffusionControlNetPipeline.from_single_file(url, controlnet=contro
 
 ## ControlNetOutput
 
-[[autodoc]] models.controlnet.ControlNetOutput
+[[autodoc]] models.controlnets.controlnet.ControlNetOutput
```
Comment (yiyixuxu) [inline docs/source/en/api/models/controlnet.md:42]: this change is not made from this PR 
it should not show up here, no?

**Comment by yiyixuxu** [comment]
can you rebase again?
I saw a bunch of changes that are not made in this PR and the commit history includes commits  that's already been merged into main (from another PR)

**Comment by MikeTkachuk** [comment]
fixed

**Review by yiyixuxu — APPROVED** [review]
thanks!
